sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,218 @@
+ * uses that buffer to set curl header. Header is set only if write operations were OK. Buffer is
+ * free after setting curl header.
+ *
+ * @param p_header
+ * @param p_headers
+ * @param separator
+ * @return az_result
+ */
+az_result az_add_header_to_curl_list(
+    az_pair * const p_header,
+    az_curl_headers_list * const p_headers,
+    az_const_span separator) {
+  // allocate a buffet for header
+  int16_t buffer_size = p_header->key.size + separator.size + p_header->value.size + 1;
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(buffer_size);
+  if (p_writable_buffer == NULL) {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  char * buffer = (char *)p_writable_buffer;
+
+  // write buffer
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = buffer_size };
+  az_result write_result = az_write_to_buffer(writable_buffer, p_header, separator);
+
+  // attach header only when write was OK
+  if (az_succeeded(write_result)) {
+    p_headers->p_list = curl_slist_append(p_headers->p_list, buffer);
+  }
+  // at any case, error or OK, free the allocated memory
+  free(p_writable_buffer);
+  return write_result;
+}
+
+/**
+ * @brief loop all the headers from a HTTP request and set each header into easy curl
+ *
+ * @param p_hrb
+ * @param p_headers
+ * @return az_result
+ */
+az_result az_build_headers(
+    az_http_request_builder * const p_hrb,
+    az_curl_headers_list * p_headers) {
+  az_const_span separator = AZ_STR(": ");
+  // get pointer to first header
+
+  az_pair a;
+  az_pair * p_header = &a;
+
+  for (uint16_t offset = 0; offset < p_hrb->headers_end; ++offset) {
+    AZ_RETURN_IF_FAILED(az_http_request_builder_get_header(p_hrb, offset, p_header));
+    AZ_RETURN_IF_FAILED(az_add_header_to_curl_list(p_header, p_headers, separator));
+  }
+
+  return AZ_OK;
+}
+
+/**
+ * @brief writes a url request adds a cero to make it a c-string. Return error if any of the write
+ * operations fails.
+ *
+ * @param writable_buffer
+ * @param url_from_request
+ * @return az_result
+ */
+az_result az_write_url(az_span const writable_buffer, az_const_span const url_from_request) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, url_from_request));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+/**
+ * handles GET request
+ */
+az_result az_curl_send_request(az_curl * const p_curl, az_http_request_builder * const p_hrb) {
+  // creates a slist for bulding curl headers
+  az_curl_headers_list headers = {
+    .p_list = NULL,
+  };
+  // build headers into a slist as curl is expecting
+  AZ_RETURN_IF_FAILED(az_build_headers(p_hrb, &headers));
+  // set all headers from slist
+  curl_easy_setopt(p_curl->p_curl, CURLOPT_HTTPHEADER, headers.p_list);
 
-az_result az_headers_to_curl(az_headers_data * const p_state, az_pair const header) {
```

Comment:
Please, leave this code. It looks like visitor patterns is much simplier and has lower probability of errors. And it can work without a builder.

Reply:
it's alive in it's own branch.  It give un-used warning that we need to work-around with (void)call() or things like that witch I don't find to be OK

Reply:
No, keep in master